Feature/simd vec3

andrearastelli opened the pull request:
Following #64

andrearastelli committed:
Merge pull request #1 from raytracingDevTeam/develop

CI for Linux and Windows

andrearastelli committed:
V1.0.0 (#38)

* Renamed Vec3 class

* Add Image

* First commit for Vec3

* Feature/image generation (#2)

* First image implementation and filesystem testing
* Update image class and removed data from vec header
* Added compile rule for C++17
* Update CMakeLists.txt
* Update .travis.yml
* Update appveyor.yml
* Update README.MD
* Image and color management classes
* Updated Image documentation
* Updated color documentation

* Updated appveyor and Vec3 fixes

* Added Ray class implementation

* Added Color constructor from Vec3

* Fixed various Vec3 bugs

Added header to safely include Vec3 in multiple files without having
multiple definitions.
Fixed variable naming in overloaded operator methods

* Updated main to generate image from rays

* Updated appveyor removing Win32 build

* Changed sqrtf to sqrt due to some strange behavior in Travis. Need investigation

* Hit sphere method for ray intersection

* Sphere color given by the sphere Normals

* Hit sphere method for ray intersection (#6)

* Sphere color given by the sphere Normals

* Defined structures for hitable components

Hitable interface.
HitableList for a container (hitable) that iterate over a list of
objects.
Sphere (Hitable) defining a hitable element that has the shape of a
sphere.

* Added new headers Hitable, HitableList, Sphere

* Operator overloading for multiplication and sum

* Removed ray_color

* Updated to render a list of objects

Instead of using <float.h>, this version uses <limits> and fetches the
MAX_FLOAT value from: std::numeric_limits<float>::max() to be C++11
compatible.

* updated to std::sqrtf

* Update .travis.yml

* Update .travis.yml

* Fixed sqrtf into sqrt

This issue only happen on Travis, using Linux.

* Feature/sphere normal (#17)

* Hit sphere method for ray intersection

* Sphere color given by the sphere Normals

* Hit sphere method for ray intersection (#6)

* Sphere color given by the sphere Normals

* Feature/antialiasing (#28)

* Added camera class

* Updated Color operators, adding += and /=

* Defined antialiasing process

Added number of samples.
Instead of using drand48(), we will use #include <random> functions, as:
std::uniform_real_distribution.

* Linearization of color

* Issue with Diffuse material

The rendered image appears to generate a "transparent" sphere.
Must be fixed.

* Fixed issue with drand48

The temporary fix uses global defined variables for generating random
values.

* Update sampling radius and fixed sphere position

The sampling for the image now is processed using a variable radius
based on the number of samples required to render the final image.

Theoretically this should be derived from all the other random function
in the engine, but for now better to keep everything simple.

* Added .gitignore

* Extended camera construction using vertical fov and posisions

* Updated scene generation to use the new camera

* Updated operation to be float

* Defined M_PI for windows

* Using cast for cos from double to float

...until windows defines cosf..

* Feature/metal material (#32)

* Added camera class

* Updated Color operators, adding += and /=

* Defined antialiasing process

Added number of samples.
Instead of using drand48(), we will use #include <random> functions, as:
std::uniform_real_distribution.

* Linearization of color

* Issue with Diffuse material

The rendered image appears to generate a "transparent" sphere.
Must be fixed.

* Fixed issue with drand48

The temporary fix uses global defined variables for generating random
values.

* Update sampling radius and fixed sphere position

The sampling for the image now is processed using a variable radius
based on the number of samples required to render the final image.

Theoretically this should be derived from all the other random function
in the engine, but for now better to keep everything simple.

* Added material classes

* Added missing file material.h

* Added script to build and run the project. Added .gitignore

* Fixed runtime error on windows

* Fixed main.cpp merge

* Feature/camera defocus (#31)

* Updated documentation

* Added static const for extracting simple data

Defined:
- X, Y, Z axis definition
- ZERO as a shorthand for Vec3(0.0f, 0.0f, 0.0f)
- ONE as a shorthand for Vec3(1.0f, 1.0f, 1.0f)

* Updated main for using the new Vec3 values

* Updated samples evaluation to 16, for testing purposes

* Fixed issues with static constexpr values

Minor other code organization for better readability.

* Updated camera with defocus.

Added random_in_unit_disc in the camera.h header file.

* Updated camera with defocus feature.

Fixed samples number to 16

* Removed Visual Studio 2015

* Feature/metal material (#32)

* Added camera class

* Updated Color operators, adding += and /=

* Defined antialiasing process

Added number of samples.
Instead of using drand48(), we will use #include <random> functions, as:
std::uniform_real_distribution.

* Linearization of color

* Issue with Diffuse material

The rendered image appears to generate a "transparent" sphere.
Must be fixed.

* Fixed issue with drand48

The temporary fix uses global defined variables for generating random
values.

* Update sampling radius and fixed sphere position

The sampling for the image now is processed using a variable radius
based on the number of samples required to render the final image.

Theoretically this should be derived from all the other random function
in the engine, but for now better to keep everything simple.

* Added material classes

* Added missing file material.h

* Added script to build and run the project. Added .gitignore

* Fixed runtime error on windows

* Fixed main.cpp merge

* Updated documentation

* Added static const for extracting simple data

Defined:
- X, Y, Z axis definition
- ZERO as a shorthand for Vec3(0.0f, 0.0f, 0.0f)
- ONE as a shorthand for Vec3(1.0f, 1.0f, 1.0f)

* Updated main for using the new Vec3 values

* Updated samples evaluation to 16, for testing purposes

* Fixed issues with static constexpr values

Minor other code organization for better readability.

* Updated camera with defocus feature.

Fixed samples number to 16

* Removed Visual Studio 2015

* Filename updated

* Fixed tabulation issue

* Now the camera uses the random variabile defined in Ray

* Moved random generation object from material to Ray.

This way the random generation object will be created in the highest
place possible.

* Added dielectric material, adjusted main.cpp to fit with the changes (#37)

* Feature/command line parameters (#34)

* Parser header, defining all the parsing functionality for command line
arguments.

To store the parameter data a struct "Parameter" is required.
Each attribute in this struct contain a single parameter data.

* Updated main with the parameter functionality.

* Updated run script with the required parameters defined.

* Updated aperture value to remove defocus

andrearastelli committed:
First SIMD setup and all inlines removed

andrearastelli committed:
First SIMD setup and all inlines removed

andrearastelli committed:
Complete Vec3 vectorization - Still working on retrieving values and setting cross product

andrearastelli committed:
Complete Vec3 vectorization - Still working on retrieving values and setting cross product

andrearastelli committed:
Updated all Vec3 to SIMD

andrearastelli committed:
Updated all Vec3 to SIMD

andrearastelli commented:
@Ale32 all Vec3 is now SIMD-ed.

It's time for testing it!!!!

andrearastelli committed:
Update system with google tests

andrearastelli committed:
Update system with google tests

andrearastelli commented:
@Ale32 Added GoogleTests package as a dependency (`ext` folder).
This way we can test any update in the system that may require some big change.

The `CMakeLists.txt` file required some updates as well, but cannot understand how the testing system can work inside the `CMake` execution.
Need to dig deeper into that.

Right now, in `CLion` I can run the tests separately from the main target, but this doesn't solve the test automation process.
Next step - will add the test automation into the `CMakeLists.txt` file so every time we send the project to travis or appvejor the testing will occur by default.

Updated `Vec3` using a union to share the memory representation used by the `float v_f[4]` and the `__m128 v` object. This way we don't need to worry about converting back and forth the data from the SIMD-like representation and the usable `float` representation.

Current Issue: if I try to render the image, I get black pixels.
Need to analyse that in more detail.
As a start I'm currently testing all the Vec3 class using gtest - because that is the only class that is changed - and try to figure out where the problem might be, in the conversion process.

andrearastelli committed:
Updated test and test scripts

andrearastelli committed:
Updated test and test scripts

andrearastelli committed:
Updated build test script

andrearastelli committed:
Updated build test script

andrearastelli committed:
Updated windows test script

andrearastelli committed:
Updated windows test script

andrearastelli commented:
@Ale32 Rebased to be on top of `develop`

andrearastelli committed:
reset test script to cmake manual build

andrearastelli committed:
reset test script to cmake manual build

andrearastelli committed:
Updated test working directory

andrearastelli committed:
Updated test working directory

Ale32 commented:
@andrearastelli cool! Thanks for your effort. Using google test is a good idea! I'll check what I get on my windows machine.

andrearastelli committed:
Updated test management and more tests for Vec3 class

andrearastelli committed:
Updated test management and more tests for Vec3 class

andrearastelli committed:
still need to figure out the dot product

andrearastelli committed:
still need to figure out the dot product

andrearastelli committed:
Test for fixing googletest on appveyor

andrearastelli committed:
Test for fixing googletest on appveyor

andrearastelli committed:
Added git submodule recursive

andrearastelli committed:
Added git submodule recursive

andrearastelli committed:
Added MSVC header specific rule

andrearastelli committed:
Added MSVC header specific rule

andrearastelli committed:
Fixed include

andrearastelli committed:
Fixed include

andrearastelli commented:
@Ale32 I've spent some time trying to figure out the linux/windows SIMD setup, but this require a more thorough test.

The gtest are correctly setup, and at this point we can safely test against the current Vec3 or the SIMD version, and with some confidence expect the same results.

andrearastelli committed:
Updated Vec3 test and SIMD implementation - Initial tests for Color

andrearastelli committed:
Updated Vec3 test and SIMD implementation - Initial tests for Color

andrearastelli committed:
SIMD working

andrearastelli committed:
SIMD working

andrearastelli committed:
SIMD render 512 samples 400x400

andrearastelli committed:
SIMD render 512 samples 400x400

andrearastelli commented:
@Ale32 before merging this, I need to figure a way to test SIMD in both travis and appveyor.
The issue with travis is that the function `_mm_hadd_ps(__m128, __m128)` is not recognized, for some reason (maybe the SIMD in travis is `SSE` and this function is available in `SSE2` or something like that).
The issues with appveyor are so many that listing those here doesn't make any sense, and I need to properly access Window and start developing there at this point, so I can commit from there the working version for appveyor.

Ale32 commented:
I'm not able to compile the code on Windows. I guess the issue is something related to this https://stackoverflow.com/questions/11228855/header-files-for-x86-simd-intrinsics#

andrearastelli committed:
Temporary removed the tests for the Windows platform due to some issue with Google Test Library

andrearastelli committed:
Temporary removed the tests for the Windows platform due to some issue with Google Test Library

andrearastelli committed:
Updated Vec3 to match windows requirements

andrearastelli committed:
Updated Vec3 to match windows requirements

andrearastelli commented:
@Ale32 ...no the issue was simpler than that...
```git submodule update --init --recursive```
If you fetch the last update from this branch you'll be able to compile everything.

The only issue I got with windows is that I cannot run the gTest suite, for some reason that is not exactly clear to me.. So, for the time being, in this branch, the `CMakeLists.txt` file has the test building disabled for windows, and will remain this way until we figure a solution to this problem.

andrearastelli committed:
Updated SIMD readme and Vec3 includes

andrearastelli committed:
Updated SIMD readme and Vec3 includes

andrearastelli committed:
SIMD-ed Color

Also: Vec3 and Color are now memory aligned to the x, y, z components
are stored one next to the other. This to avoid possible SIMD issues.

andrearastelli committed:
SIMD-ed Color

Also: Vec3 and Color are now memory aligned to the x, y, z components
are stored one next to the other. This to avoid possible SIMD issues.

andrearastelli commented on src/vec3.h:
```diff
@@ -5,12 +5,20 @@
+#endif
 
-class Vec3
+class alignas(16) Vec3
```
@Ale32 The previous definition for memory aligned class arguments was very old and not supported anymore.
This one should work for windows as well.

andrearastelli commented:
@Ale32 Color is SIMD-ed as well. This let us gain another 2% of performance.

Now some proper profiling is required to avoid random number in performance improvements.

andrearastelli commented:
Task required in order to properly continue working on this PR:
- [ ] Make the tests work on `travis` and `appveyor`.
- [ ] Make the `gTest` suite work on both Linux and Windows.
- [ ] Profile the SIMD code against the old one, to have a proper performance table.

andrearastelli committed:
SIMD hardware checked

from https://stackoverflow.com/a/7495023/1162467

andrearastelli committed:
SIMD hardware checked

from https://stackoverflow.com/a/7495023/1162467

andrearastelli committed:
Added Vec3 SIMD version checker for SSE3

andrearastelli committed:
Added Vec3 SIMD version checker for SSE3

andrearastelli committed:
test execution

andrearastelli committed:
test execution

andrearastelli committed:
Debug test folder structure

andrearastelli committed:
Debug test folder structure

andrearastelli committed:
Installed tree

andrearastelli committed:
Installed tree

andrearastelli committed:
Find test executable files in the project dir

andrearastelli committed:
Find test executable files in the project dir

andrearastelli committed:
Print test location

andrearastelli committed:
Print test location

andrearastelli committed:
Removed parallel hadd

andrearastelli committed:
Removed parallel hadd

andrearastelli committed:
Fixed hadd issue in Vec3

andrearastelli committed:
Fixed hadd issue in Vec3

andrearastelli committed:
Reset test system

andrearastelli committed:
Reset test system

andrearastelli committed:
Updated tests and listing system

andrearastelli committed:
Updated tests and listing system

andrearastelli commented:
@Ale32 Fixed travis testing setup. Now I need to figure how to not execute the `_mm_hadd_ps` when the hardware doesn't support it.

andrearastelli committed:
Added windows headers

andrearastelli committed:
Added windows headers

andrearastelli committed:
Updated from unsigned to int

andrearastelli committed:
Updated from unsigned to int

andrearastelli committed:
Updated SIMD check class

andrearastelli committed:
Updated SIMD check class

andrearastelli committed:
Added temporary SIMD checking system

andrearastelli committed:
Added temporary SIMD checking system

andrearastelli committed:
Reorganised travis install scripts

andrearastelli committed:
Reorganised travis install scripts

andrearastelli committed:
Added SIMD check script

Can be useful to test if the current machine has a very specific SIMD
hardware implementation.

andrearastelli committed:
Added SIMD check script

Can be useful to test if the current machine has a very specific SIMD
hardware implementation.

andrearastelli committed:
Added gates over _mm_hadd_ps because only available in SSE3

andrearastelli committed:
Added gates over _mm_hadd_ps because only available in SSE3

andrearastelli committed:
Various fixes to the main function and better management of memory in CornellBox function

andrearastelli committed:
Various fixes to the main function and better management of memory in CornellBox function

andrearastelli commented:
This PR is starting to be **huge**, compared to the other ones.

Anyway.. the stuff that is updated, so far:
# Vec3 now runs with parallel code
# Color now runs with parallel code
# `simd.h` added. Because some of the features of Vec3 and Color requires horizontal sums, and only starting from SSE3 and forward those implementation actually take place, like `_mm_hadd_ps`, a new SIMD check system needed to be designed
# For reference about SIMD on intel platform: https://software.intel.com/sites/landingpage/IntrinsicsGuide/#expand=2777,2777&text=_mm_hadd_ps
# Various performance and memory check helped me figure that the main function could have run with far less memory allocated - so fixed that
# To test the SIMD implementation the `gTest` package needed to be added
# Various test implemented for `Vec3`
# Various test implemented for `Color`
# The `.travis.yml` script required a big update to support test cases
# The `appveyor.yml` script required some updates to support test cases (but still have lots of issues)

Stuff that is still not finished:
- [ ] Have `appveyor.yml` working and testing the proper code on windows
- [ ] Have the windows implementation work the same as the Linux one

@Ale32 This is the current state of the art of this PR.

andrearastelli committed:
Unfortunately, to use the Windows.h header, some macro has to be undefined@

andrearastelli committed:
Unfortunately, to use the Windows.h header, some macro has to be undefined@

andrearastelli committed:
Updated test build for windows

andrearastelli committed:
Updated test build for windows

andrearastelli committed:
Removed hadd and updated appveyor build script

andrearastelli committed:
Removed hadd and updated appveyor build script

andrearastelli committed:
Removed x64 configuration

andrearastelli committed:
Removed x64 configuration

andrearastelli committed:
Fixed dot product and length evaluation

andrearastelli committed:
Fixed dot product and length evaluation

andrearastelli committed:
removed static library build

andrearastelli committed:
removed static library build

andrearastelli committed:
Build gtest as shared lib

andrearastelli committed:
Build gtest as shared lib

andrearastelli committed:
Test relocation

andrearastelli committed:
Test relocation

andrearastelli committed:
Updated test with debug configuration

andrearastelli committed:
Updated test with debug configuration

andrearastelli committed:
Updates on appveyor build script

andrearastelli committed:
Updates on appveyor build script

andrearastelli committed:
Updated folder lookup

andrearastelli committed:
Updated folder lookup

andrearastelli commented:
@Ale32 ALL TEST PASSED SUCCESSFULLY!!!!!!!

andrearastelli committed:
Merge branch 'master' into develop

andrearastelli committed:
Merge branch 'feature/SIMD-vec3' of github.com:raytracingDevTeam/raytracing into feature/SIMD-vec3

andrearastelli commented:
Rebased to be on top of develop.

Ale32 commented on src/color.h:
```diff
@@ -38,7 +47,7 @@ class Color
      * @param b Blue component.
      */
-    Color(int r, int g, int b): c{r / 255.0f, g / 255.0f, b / 255.0f} {}
+    Color(int r, int g, int b, int alpha=255): v{_mm_set_ps(alpha/255.0f, b/255.0f, g/255.0f, r/255.0f)} {}
```
Why (alpha, b, g, r)?

Ale32 commented on src/hitable.h:
```diff
@@ -132,6 +132,7 @@ RotateY::RotateY(Hitable *p, float angle) : ptr(p)
         for (auto j=0; j<2; ++j)
             for (auto k=0; k<2; ++k)
             {
+                // TOODO: this can be SIMDed - needs more check.
```
Typo?
